DOCSP-24542 Authentication Mechanisms (#38)
mongodb/docs-csharp: source/fundamentals/authentication.txt

.. _csharp-authentication-mechanisms:

=========================
Authentication Mechanisms
=========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

> In this guide, you can learn how to authenticate with MongoDB using each
> **authentication mechanism** available in the {+mongo-community+}.

Reviewer: ```suggestion
In this guide, you can learn how to authenticate with MongoDB using the
**authentication mechanisms** available in the {+mongo-community+}.
```

Authentication mechanisms are processes by which the driver and server confirm

> identity and establish trust to ensure security before connecting.

Reviewer: The wording here is a little unclear. confirm whose identity? Establish trust between which entities?

Author: Reworded this to hopefully make it more clear

You can use the following authentication mechanisms with the latest version
of {+mongo-community+}:

- :ref:`csharp-scram-sha-256`
- :ref:`csharp-scram-sha-1`
- :ref:`csharp-x509`

To authenticate using ``Kerberos`` or ``LDAP``, see the

> :ref:`csharp-enterprise-authentication-mechanisms` guide. For more information

Reviewer: this ref is showing up empty in the build. If the page isnt populated yet, maybe add a title to the shell page? or leave this out for now.

Author: Added a title to the shell page for now

on establishing a connection to your MongoDB cluster, see the
:ref:`csharp-connection` guide.

Specify an Authentication Mechanism
-----------------------------------

You can specify your authentication mechanism and credentials when connecting to

> MongoDB using either of the following:

Reviewer: ```suggestion
MongoDB using either of the following methods:
```

- A connection string
- A ``MongoCredential`` factory method

> A **connection string**, also known as a **connection uri**, is a string that

Reviewer: ```suggestion
A **connection string**, also known as a **connection URI**, is a string that
```

tells the driver how to connect to a MongoDB deployment and how to behave while
connected.

The ``MongoCredential`` class contains factory methods for each of the supported

> authentication mechanisms. 

Reviewer: maybe you can put these descriptions in the list

Mechanisms
----------

The following examples specify authentication mechanisms using the following
placeholders:

- ``username`` - Your MongoDB username
- ``password`` - Your MongoDB user's password
- ``hostname`` - The network address of your MongoDB server, accessible by your client
- ``port`` - The port number of your MongoDB server
- ``authenticationDb`` - The MongoDB database that contains your user's authentication
  data. If you omit this parameter, the driver uses the default value ``admin``.

Select the :guilabel:`Connection String` or :guilabel:`MongoCredential` tab to

> see the corresponding syntax for specifying your authentication mechanism.

Reviewer: After reading this, I expected to see the tabs right below it but there was another section.

Reviewer: But then this same text is replicated in the Default section -- should this be here?

Author: I initially intended it to just be a statement that applies to the whole page. But I agree it's a little confusing here. Removed it from here and left it in the other areas

   supports ``MONGODB-CR``. These versions automatically determine which
   version of SCRAM to use.

Select the :guilabel:`Connection String` or :guilabel:`MongoCredential` tab to
see the corresponding syntax for specifying the default authentication mechanism:

.. tabs::

   .. tab:: Connection String
      :tabid: default-connection-string

      .. code-block:: csharp

         var mongoClient = new MongoClient("mongodb://<username>:<password>@<hostname>:<port>/?authSource=<authenticationDb>");

   .. tab:: MongoCredential
      :tabid: default-mongo-credential

      .. code-block:: csharp

         var credentials = MongoCredential.CreateCredential("<authenticationDb>", "<username>", "<password>");
         var settings = MongoClientSettings.FromConnectionString("<host>");
         settings.Credential = credentials;
         var mongoClient = new MongoClient(settings);

.. _csharp-scram-sha-256:

SCRAM-SHA-256
~~~~~~~~~~~~~

``SCRAM-SHA-256`` is a salted challenge-response authentication mechanism (SCRAM)
that uses your username and password, encrypted with the ``SHA-256`` algorithm,
to authenticate your user. 

.. tip:: Default Mechanism

>    MongoDB version 4.0 and later uses SCRAM as the default authentication mechanism.

Reviewer: which SCRAM method? It's a little confusing that this same tip is in both this section and the next.

Author: I updated the tip to make it more clear.

   To learn more on specifying the default mechanism, see :ref:`csharp-authentication-default`.

.. _csharp-scram-sha-1:

SCRAM-SHA-1
~~~~~~~~~~~

``SCRAM-SHA-1`` is s a salted challenge-response mechanism (SCRAM) that uses 
your username and password, encrypted with the ``SHA-1`` algorithm, to authenticate 
your user.

.. tip:: Default Mechanism

   MongoDB version 4.0 and later uses SCRAM as the default authentication mechanism.
   To learn more on specifying the default mechanism, see :ref:`csharp-authentication-default`.

.. _csharp-x509:

X.509
~~~~~

> The ``X.509`` authentication mechanism uses `TLS <https://en.wikipedia.org/wiki/Transport_Layer_Security>`__

Reviewer: there is a wikipedia reference:

```suggestion
The ``X.509`` authentication mechanism uses :wikipedia:`TLS </Transport_Layer_Security>`
```

Author: I didn't know about that reference, thanks!

with ``X.509`` certificates to authenticate your user, identified by the

> distinguished names of your client certificate. When you specify the 

Reviewer: what does "distinguished names of your client certificate" mean? Or should we expect readers to understand?

Author: I think defining that is out of scope for this page. If the user is planning on using X509 I think they would have an idea on what this means, and they can read more information on it in the X.509 guide linked below.

``X.509`` authentication mechanism, the server authenticates the connection using 
the subject name of the client certificate.

To learn more about using TLS/SSL, see our :ref:`TLS/SSL guide <csharp-tls>`.

To learn more about ``X.509`` certificates, see the :ref:`X.509 guide <x509-client-authentication>`.

Select the :guilabel:`Connection String` or :guilabel:`MongoCredential` tab to
see the corresponding syntax for specifying the ``X.509`` authentication mechanism:

.. tabs::

   .. tab:: Connection String
      :tabid: default-connection-string

      .. code-block:: csharp

>          var connectionString = "mongodb://"<hostname>"/?authMechanism=MONGODB-X509";

Reviewer: this string is rendering weirdly with the codeblock formatting. Can you remove the quotes around `<hostname>`?
```suggestion
         var connectionString = "mongodb://<hostname>/?authMechanism=MONGODB-X509";
```

         var settings = MongoClientSettings.FromConnectionString(connectionString);

         settings.useTls = true;
         settings.SslSettings = new SslSettings
         {
            ClientCertificates = new List<X509Certificate>()
            {

>                 new X509Certificate2("<path to X.509 certificate>", "<X.509 certificate password>")

Reviewer: same thing here-- if the <> phrases are placeholders, do you need the quotes?

Author: In this case I think the quotes are helpful as they tell the user that the placeholder is meant to be a string value rather than some kind of object or something. Since it doesn't seem to be effecting the rendering in this case I'd vote to leave them in here.

            }
         };

   .. tab:: MongoCredential
      :tabid: default-mongo-credential

      .. code-block:: csharp

>          var credentials = MongoCredential.CreateMongoX509Credential("<X.509 certificate username>")

Reviewer: same comments apply about the quotes and placeholders in this codeblock

Author: See above response

.. _csharp-authentication-mechanisms:

=========================
Authentication Mechanisms
=========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to authenticate with MongoDB using the
**authentication mechanisms** available in the {+mongo-community+}.
Authentication mechanisms are processes by which the driver and server confirm
the identity of a client to ensure security before connecting.

You can use the following authentication mechanisms with the latest version
of {+mongo-community+}:

- :ref:`csharp-scram-sha-256`
- :ref:`csharp-scram-sha-1`
- :ref:`csharp-x509`

To authenticate using ``Kerberos`` or ``LDAP``, see the

> :ref:`csharp-enterprise-authentication-mechanisms` guide. For more information

Author: Yep, this is the next page on my backlog

on establishing a connection to your MongoDB cluster, see the
:ref:`csharp-connection` guide.

Specify an Authentication Mechanism
-----------------------------------

You can specify your authentication mechanism and credentials when connecting to
MongoDB using either of the following methods:

- A **connection string**, also known as a **connection URI**, which is a string
  that tells the driver how to connect to a MongoDB deployment and how to behave while
  connected.

- A factory method for the supported authentication mechanism, contained in the 
  ``MongoCredential`` class.

Mechanisms
----------

The following examples specify authentication mechanisms using the following
placeholders:

- ``username`` - Your MongoDB username
- ``password`` - Your MongoDB user's password
- ``hostname`` - The network address of your MongoDB server, accessible by your client
- ``port`` - The port number of your MongoDB server

> - ``authenticationDb`` - The MongoDB database that contains your user's authentication

Author: Good idea! Added
